### DISCUSSION TITLE
Padding problem with Matplotlib 3D plot in VSCode

### DISCUSSION THREAD

**Discussion by tueda** [microsoft/vscode-jupyter#9132] [Questions and Answers] [answered]
I encountered a padding problem in a Matplotlib 3D plot, where the exponent to the right side of the z-axis was truncated in VSCode, making it very confusing at first glance.

As I write below, I can avoid it by changing viewport configurations. But, I wonder if
- this may be a VSCode issue; the generated picture is square compared with those created in other Jupyter environments, which leads to truncation of the exponent. Some default viewport configurations might not be so adequate,
- this may be a Matplotlib issue; the bounding box doesn't seem to be calculated correctly, or
- this may be an issue in each environment; I should change some configurations to customize the default viewport.

### Code

```python
import matplotlib.pyplot as plt
import numpy as np

x = np.linspace(0, 1)
y = np.linspace(0, 1)
xx, yy = np.meshgrid(x, y)
zz = (xx ** 2 + yy ** 2) * 1e-5

fig, ax = plt.subplots(subplot_kw=dict(projection='3d'))
ax.plot_surface(xx, yy, zz)
fig.tight_layout()
```

### Result in VSCode

![image](https://user-images.githubusercontent.com/1481665/155835381-9eb2ecdd-c554-4242-b7eb-3a0a110fabd3.png)

Moreover, clicking the "Save As" button in the output cell gave me `output.png`, which was truncated more aggressively:

![image](https://user-images.githubusercontent.com/1481665/155835940-9b1fc420-ae5d-4da3-bc5d-04c9cd60054d.png)

```
Version: 1.64.2 (system setup)
Commit: f80445acd5a3dadef24aa209168452a3d97cc326
Date: 2022-02-09T22:02:28.252Z
Electron: 13.5.2
Chromium: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.19044

Python 3.10.2
Matplotlib 3.5.1
Ubuntu 20.04.3 (WSL2)
```

### What I expected

Here is the result in Google Colaboratory:
![image](https://user-images.githubusercontent.com/1481665/155835602-c2a4b8b0-f180-4cd7-8e40-d8185de2ea4d.png)
I can see the order of magnitude of the z-axis.

### Workaround

Setting a space as the z-label fixes the problem.
```python
ax.set_zlabel(' ')
```
Or, I can adjust the "distance":
```python
ax.dist = 15
```

**Comment by DonJayamanne** [comment, marked as answer]
Thanks for filing this here.
I've tried this outside of VS Code (using classic Jupyter Notebook) & I'm experiencing the exact same issue.
I.e. this is the way the image is rendered & saved by Jupyter as well.

Here's a screenshot of the png file I downloaded & saved from Jupyter Notebooks (i.e. outside VS Code)
![image](https://user-images.githubusercontent.com/1948812/155922024-eb78d1be-2b35-431c-9263-3ee51a5dfaf2.png)

I believe the screenshot of Google Colaboratory is just a rendering of the output, i.e. it works the same way as it works VS Code & Jupyter Notebooks, & might not work when you save it (i.e. it will be clipped when saving from Google Colab).

As this is a bug in Matplot lib library itself and I'd reccomend filing a bug against their repo.
Looking at their docs, it might be a known issue, https://matplotlib.org/stable/tutorials/intermediate/tight_layout_guide.html#caveats

**Reply by tueda** [reply]
Thank you for your reply and for testing the problem.

Yes, Jupyter Notebooks seem to give the same picture... It was my fault not to check it on pure Jupyter Notebooks, sorry.

Though now I am starting to wonder if this is a bug in Matplotlib or that in Jupyter itself (not VSCode Jupyter extension), anyway, I agree that this is not a VSCode issue.
